tools/istio-docker.mk

```diff
@@ -21,7 +21,21 @@
 # It does not upload to a registry.
 docker: build-linux test-bins-linux docker.all
 
-DOCKER_TARGETS:=docker.pilot docker.proxy_debug docker.proxytproxy docker.proxyv2 docker.app docker.app_sidecar docker.test_policybackend \
+# The CI system changes VERSION from the top level makefile per build. As a result,
+# when running in CI, we always want docker.base to be generated and tagged. If the
+# base image is not regenerated and tagged per release, the tag used is undefined and
+# the Istio build is not reproducible.
+#
+# DO NOT ALTER this block without understanding how it works. Please consult #test-and-release.
+ifeq (${CI},"prow")
```

Comment:
Why do we need this? Is it just to avoid rebuilding locally when you do `make docker`? If so it will only take like 1s once its cached I think?

It seems like a bad pattern to check if we are in prow

Reply:
Also when do we actually want to build a new base image? In all the e2e/integration tests it is using 1.3-dev, but the release test is not. is it only when we release we want it?

Reply:
> Why do we need this? Is it just to avoid rebuilding locally when you do `make docker`? If so it will only take like 1s once its cached I think?
> 
> It seems like a bad pattern to check if we are in prow

This is needed to avoid rebuilding locally.  The base image is built with `--no-cache` which basically means "wipe out what you got and build afresh".  This takes 30 seconds on my very fast gear, possibly longer on slower gear.

I agree in principle checking environment variables to conditionalize default build sematics is not ideal.  A possibly better approach is to run "make base" prior to the rest of the make.

Unfortunately the make command is tied up in that awful `udpateVersions.sh` script - and recurive makefile executions of that from the top level makefile.  SO.............  Feeling a little overloaded on how to get this work in.  I think the prow conditional could be removed if we could remove `updateVersions.sh`.  Unfortunately we have been **attemping** to remove this script for over 18 months with little success.

Reply:
> Also when do we actually want to build a new base image? In all the e2e/integration tests it is using 1.3-dev, but the release test is not. is it only when we release we want it?

great question and an area I could use help with.  We want to build a new base image during the following events:
- gate checks
- release process